File: bindings/dotnet/src/Zlink/Runtime/Options/ContextOptions.cs
// SPDX-License-Identifier: MPL-2.0

using System.Text;
using Systems.Zlink.Runtime.Native;

namespace Systems.Zlink;

internal sealed class ContextOptions : IContextOptions
{
    private readonly Context _context;
    private string _threadNamePrefix = string.Empty;

    internal ContextOptions(Context context)
    {
        _context = context;
    }

    public int IoThreads
    {
        get => _context.GetOption(ContextOption.IoThreads);
        set => _context.SetOption(ContextOption.IoThreads, value);
    }

    public int MaxSockets
    {
        get => _context.GetOption(ContextOption.MaxSockets);
        set => _context.SetOption(ContextOption.MaxSockets, value);
    }

    public int SocketLimit => _context.GetOption(ContextOption.SocketLimit);

    public int ThreadPriority
    {
        get => _context.GetOption(ContextOption.ThreadPriority);
        set => _context.SetOption(ContextOption.ThreadPriority, value);
    }

    public int ThreadSchedulingPolicy
    {
        get => _context.GetOption(ContextOption.ThreadSchedPolicy);
        set => _context.SetOption(ContextOption.ThreadSchedPolicy, value);
    }

    public int MaxMessageSize
    {
        get => _context.GetOption(ContextOption.MaxMsgSz);
        set => _context.SetOption(ContextOption.MaxMsgSz, value);
    }

    public int MessageThreadSize => _context.GetOption(ContextOption.MsgTSize);

    public bool Blocky
    {
        get => _context.GetOption(ContextOption.Blocky) != 0;
        set => _context.SetOption(ContextOption.Blocky, value ? 1 : 0);
    }

    public AutoHwmProfile AutoHwmProfile
    {
        get => (AutoHwmProfile)_context.GetOption(ContextOption.AutoHwmProfile);
        set => _context.SetOption(ContextOption.AutoHwmProfile, (int)value);
    }

    public ulong AutoHwmMessageUnitBytes
    {
        get => _context.GetUInt64Option(ContextOption.AutoHwmMsgUnitBytes);
        set => _context.SetUInt64Option(ContextOption.AutoHwmMsgUnitBytes, value);
    }

    public bool AutoHwmEnabled
    {
        get => _context.GetOption(ContextOption.AutoHwmEnabled) != 0;
        set => _context.SetOption(ContextOption.AutoHwmEnabled, value ? 1 : 0);
    }

    public TimeSpan AutoHwmRecalcDebounce
    {
        get => TimeSpan.FromMilliseconds(
            _context.GetOption(ContextOption.AutoHwmRecalcDebounce));
        set => _context.SetOption(ContextOption.AutoHwmRecalcDebounce,
            EncodeMilliseconds(value, nameof(value)));
    }

    public string ThreadNamePrefix
    {
        get => _threadNamePrefix;
        set
        {
            if (value == null)
                throw new ArgumentNullException(nameof(value));
            if (value.IndexOf('\0') >= 0)
                throw new ArgumentException("Value must not contain NUL.",
                    nameof(value));
            if (value.Length > 16)
                throw new ArgumentOutOfRangeException(nameof(value));
            var bytes = Encoding.UTF8.GetBytes(value);
            if (bytes.Length > 16)
                throw new ArgumentOutOfRangeException(nameof(value));
            var rc = NativeMethods.zlink_ctx_set_data(_context.Handle,
                (int)ContextOption.ThreadNamePrefix, bytes, (nuint)bytes.Length);
            ZlinkException.ThrowConfigIfError(rc);
            _threadNamePrefix = value;
        }
    }

    public void AddThreadAffinityCpu(int cpu)
    {
        _context.SetOption(ContextOption.ThreadAffinityCpuAdd, cpu);
    }

    public void RemoveThreadAffinityCpu(int cpu)
    {
        _context.SetOption(ContextOption.ThreadAffinityCpuRemove, cpu);
    }

    private static int EncodeMilliseconds(TimeSpan value, string paramName)
    {
        var millis = value.TotalMilliseconds;
        if (double.IsNaN(millis) || double.IsInfinity(millis)
                                 || millis < 0 || millis > int.MaxValue)
            throw new ArgumentOutOfRangeException(paramName);
        return (int)Math.Ceiling(millis);
    }
}

File: bindings/java/src/main/java/systems/zlink/contracts/core/ContextOptions.java
/* SPDX-License-Identifier: MPL-2.0 */

package systems.zlink.contracts.core;

import systems.zlink.contracts.sockets.AutoHwmProfile;
import systems.zlink.internal.ContractAccess;
import systems.zlink.internal.DurationConversions;
import java.time.Duration;
import java.util.Objects;

/**
 * Typed facade for context options.
 */
public final class ContextOptions {
    private final Context context;
    private String threadNamePrefix = "";

    public ContextOptions(Context context) {
        this.context = context;
    }

    public int ioThreads() {
        return ContractAccess.contextGetOption(context, ContextOption.IO_THREADS);
    }

    public void ioThreads(int count) {
        ContractAccess.contextSetOption(context, ContextOption.IO_THREADS, count);
    }

    public int maxSockets() {
        return ContractAccess.contextGetOption(context, ContextOption.MAX_SOCKETS);
    }

    public void maxSockets(int count) {
        ContractAccess.contextSetOption(context, ContextOption.MAX_SOCKETS, count);
    }

    public int socketLimit() {
        return ContractAccess.contextGetOption(context, ContextOption.SOCKET_LIMIT);
    }

    public int threadPriority() {
        return ContractAccess.contextGetOption(context, ContextOption.THREAD_PRIORITY);
    }

    public void threadPriority(int priority) {
        ContractAccess.contextSetOption(context, ContextOption.THREAD_PRIORITY, priority);
    }

    public int threadSchedulingPolicy() {
        return ContractAccess.contextGetOption(context, ContextOption.THREAD_SCHED_POLICY);
    }

    public void threadSchedulingPolicy(int policy) {
        ContractAccess.contextSetOption(context, ContextOption.THREAD_SCHED_POLICY, policy);
    }

    public String threadNamePrefix() {
        return threadNamePrefix;
    }

    public void threadNamePrefix(String prefix) {
        Objects.requireNonNull(prefix, "prefix");
        ContractAccess.contextSetOptionData(context,
            ContextOption.THREAD_NAME_PREFIX, prefix);
        threadNamePrefix = prefix;
    }

    public int maxMessageSize() {
        return ContractAccess.contextGetOption(context, ContextOption.MAX_MSGSZ);
    }

    public void maxMessageSize(int bytes) {
        ContractAccess.contextSetOption(context, ContextOption.MAX_MSGSZ, bytes);
    }

    public int messageThreadSize() {
        return ContractAccess.contextGetOption(context, ContextOption.MSG_T_SIZE);
    }

    public boolean blocky() {
        return ContractAccess.contextGetOption(context, ContextOption.BLOCKY) != 0;
    }

    public void blocky(boolean enabled) {
        ContractAccess.contextSetOption(context, ContextOption.BLOCKY,
            enabled ? 1 : 0);
    }

    public boolean autoHwmEnabled() {
        return ContractAccess.contextGetOption(context,
            ContextOption.AUTO_HWM_ENABLE) != 0;
    }

    public void autoHwmEnabled(boolean enabled) {
        ContractAccess.contextSetOption(context, ContextOption.AUTO_HWM_ENABLE,
            enabled ? 1 : 0);
    }

    public Duration autoHwmRecalcDebounce() {
        return Duration.ofMillis(
          ContractAccess.contextGetOption(context,
              ContextOption.AUTO_HWM_RECALC_DEBOUNCE_MS));
    }

    public void autoHwmRecalcDebounce(Duration value) {
        Objects.requireNonNull(value, "value");
        ContractAccess.contextSetOption(context,
          ContextOption.AUTO_HWM_RECALC_DEBOUNCE_MS,
          DurationConversions.toIntMillis(value, "value"));
    }

    public AutoHwmProfile coreHwmProfile() {
        return AutoHwmProfile.fromValue(
          ContractAccess.contextGetOption(context, ContextOption.AUTO_HWM_PROFILE));
    }

    public void coreHwmProfile(AutoHwmProfile profile) {
        ContractAccess.contextSetOption(context, ContextOption.AUTO_HWM_PROFILE,
            profile.value());
    }

    /**
     * Returns the explicit context memory limit in bytes.
     */
    public long coreHwmMemoryLimitBytes() {
        return ContractAccess.contextGetUInt64Option(context,
            ContextOption.AUTO_HWM_MEMORY_LIMIT_BYTES);
    }

    /**
     * Sets the explicit context memory limit in the positive Java long range.
     */
    public void coreHwmMemoryLimitBytes(long value) {
        requireNonNegativeLong(value);
        ContractAccess.contextSetUInt64Option(context,
            ContextOption.AUTO_HWM_MEMORY_LIMIT_BYTES, value);
    }

    /** Returns the manually configured context-wide Core budget in bytes. */
    public long coreHwmBudgetBytes() {
        return ContractAccess.contextGetUInt64Option(context,
            ContextOption.AUTO_HWM_CORE_BUDGET_BYTES);
    }

    /** Sets the context-wide Core budget in the positive Java long range. */
    public void coreHwmBudgetBytes(long value) {
        requireNonNegativeLong(value);
        ContractAccess.contextSetUInt64Option(context,
            ContextOption.AUTO_HWM_CORE_BUDGET_BYTES, value);
    }

    public void addThreadAffinityCpu(int cpu) {
        ContractAccess.contextSetOption(context,
            ContextOption.THREAD_AFFINITY_CPU_ADD, cpu);
    }

    public void removeThreadAffinityCpu(int cpu) {
        ContractAccess.contextSetOption(context,
            ContextOption.THREAD_AFFINITY_CPU_REMOVE, cpu);
    }

    private static void requireNonNegativeLong(long value) {
        if (value < 0) {
            throw new IllegalArgumentException(
                "byte value exceeds Java's positive long range");
        }
    }

}
